### PR TITLE
Time partitioning

### PR DESCRIPTION
**Pull request by mubeen**
Added support for the MONTH and YEAR time partitioning capabilities of Big Query as well as upgraded the following google libraries:

-  google.auth = 0.21.1
-  google.cloud = 1.119.0
-  google.cloud.storage = 1.113.4

### CODE REVIEW

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:441]: I think we can leave this line as-is; with intermediate tables the exact partitioning type shouldn't matter too much. As the comment above details, the only reason we want to partition the table to begin with is to use the `__PARTITIONTIME` pseudocolumn to avoid trying to delete rows that are still in the streaming buffer.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:472]: This method can be called in both the case of creating a table and updating one, correct? At least in the first case, does it make sense to even try to perform the lookup here? And if we're updating a table, do we have to explicitly set a time partitioning type? Have you tried leaving the partitioning type unset entirely and seeing if the existing partitioning type remains in use?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/utils/PartitionedTableId.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/utils/PartitionedTableId.java:248]: Any BigQuery docs you can reference here on this syntax?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/utils/PartitionedTableId.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/utils/PartitionedTableId.java:252]: Any BigQuery docs you can reference here on this syntax?

File: pom.xml
Comment (C0urante) [inline pom.xml:40]: Can we keep the indentation the same here?

**Review by C0urante — COMMENTED** [review]
Thanks @mubeen! I've given this a first pass. Have only looked at the functional changes; haven't had time for the test code yet.

I also have some high-level thoughts about how you envision this feature working--I was imagining a new configuration property that would dictate the partitioning type the connector uses for every table, and if users wanted different partitioning types for different tables, they could bring up different connectors. It seems like you're trying to go for a no-configuration-required approach where the partitioning type is automatically deduced by the connector, which is probably a better user experience. However, this leaves two questions:

1. Should we allow users to configure the partitioning type for tables that the connector creates automatically?
2. In order to prevent excessive table metadata requests to BigQuery, can we perform some caching of table partitioning type so that it only has to be looked up once per task per table? It's possible that a table might be deleted and then recreated with a different partitioning type over the lifetime of the connector, but it's such a rare case with such an easy remediation (just restart the connector and it'll read the new partitioning type) that I'm not sure we'd need to spend a lot of time/effort guarding against that case besides maybe a friendly log message with instructions on what happened and how the user can proceed.

File: pom.xml
Comment (mubeen) [reply pom.xml:154]: Hi, the latest available version for google cloud storage is 1.113.4 at the time when this work was done. The version of the google cloud bigquery artefact is 1.119.0, hence the need to have diverging versions

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (mubeen) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:22]: Will resolve

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (mubeen) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:453]: Will resolve

**Comment by robertotena** [comment]
One question! Any reason for not implementing also the `HOUR` partitioning too?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/utils/PartitionedTableId.java
Comment (robertotena) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/utils/PartitionedTableId.java:248]: I assume is this one:
https://cloud.google.com/bigquery/docs/partitioned-tables

image

File: kcbq-connector/pom.xml
Comment (robertotena) [inline kcbq-connector/pom.xml:71]: Do we need a cache? I was checking the SchemaManager, and I don't have the impression we are going to have such an amount of BigQuery metadata instances to cause a memory issue?

If that's the case, I'd rather use just a `HashMap` instead of a library.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (robertotena) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:24]: I'd keep it consistent and not collapse the imports

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (robertotena) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:457]: It feels odd to have a `setPartitioningForDefinition` table that sets the `TimePartitioning` [here](https://github.com/confluentinc/kafka-connect-bigquery/pull/61/files#diff-ca041c1c31ebc1ccb34783c9f1a727c172e94a985a0d82ec6ba39c3d487b73a4R488), returns the value and sets it here again. I'd rather rename the other method `getPartitioningForDefinition` that only returns the value, and then here you decide how you want to set it.

**Comment by C0urante** [comment]
@robertotena thanks for the additional review, I agree with your comments (at least, the ones that I've read by the time of posting this).

@mubeen I've been holding off on another round of review on the assumption that you'd ping me when you were ready; if this isn't the case and you've been ready for another round let me know and I can take another look.

**Comment by mubeen** [comment]
Hi Guys,

Thanks for the feedback, I will give it one more go to include the HOURLY and make adjustments as recommended and ping you guys to do a final review.

Thanks for the assistance.

File: kcbq-connector/pom.xml
Comment (mubeen) [reply kcbq-connector/pom.xml:71]: Hi @robertotena,

@C0urante suggested that we cache the tables to prevent a large number of requests to BigQuery.

The reason I opted for a library is to allow configuration for the size of the cache and a cache expiry per entry. Currently I have set it so that you can configure against the number of hours that you would like an entry to be held within the cache. We can expand on this configuration to allow users to choose against the enumeration values which are present in java.util.concurrent.TimeUnit.

**Comment by mubeen** [comment]
@C0urante;

Re-opened pull request 61 and closed pull request 62 as requested. That was an error on my side.

Thanks

File: kcbq-connector/pom.xml
Comment (C0urante) [reply kcbq-connector/pom.xml:71]: I don't see an advantage to expiring entries from the cache, considering that the partitioning type for a table can't be modified after it's created, and it would take a gigantic number of tables in the cache to incur a non-negligible performance hit. I think a `HashMap` is fine here too 👍

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java]: I think this still needs to be addressed? Line numbers have changed from 452-453 to 468-469.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:362]: Discussed elsewhere, just want to comment here for the record--I think the logic here is fine but we should replace use of the third-party `Cache` class with a simple `Map<String, Table>`.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:157]: If the cache is going to be accessed from the `SchemaManager`, we need to either make it thread-safe, or manually synchronize access to it, since the `SchemaManager` and the `BigQuerySinkTask` might access it at the same time.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java]: I think this still needs to be addressed? Specifically:

- If a table is being created instead of updated, we shouldn't look up the partitioning type for the table since it shouldn't exist in the first place (of course, there's always the possibility of a race condition with other tasks and/or manual actions on the part of the user, but a lookup here won't prevent those in the case that the other tasks or the user act after the lookup has been performed).
- If a table is being updated instead of created, have you tried leaving the partitioning type unset entirely and seeing if the existing partitioning type remains in use?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/utils/PartitionedTableId.java
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/utils/PartitionedTableId.java:268]: 👍    thanks @robertotena

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkTaskConfig.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkTaskConfig.java:273]: Discussed elsewhere, commenting here for the record--don't think we want to add these configurable properties unless we opt to use an expirable cache instead of a simple hash map.

**Comment by C0urante** [comment]
Thanks @mubeen, I've done the next round of review. There's also the high-level question I left in my last round that has yet to be addressed:

> Should we allow users to configure the partitioning type for tables that the connector creates automatically?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (mubeen) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java]: This method has been renamed to getTimePartitioningForDefinition(tableId), its current usage is only from within the constructTableInfo method. In the original code the time partitioning was being hard-coded to DAY. As requested for the intermediateTables I have reverted it to DAY, but for the non intermediate tables I felt that we should do the look up and if none is found it will be set to DAY as the default as was in the original codebase.

The returned time partitioning is utilised to set the timestampPartitionFieldName field and in the case of clusteringFieldName field if clustering is present. If I were to remove this method, how do you propose we go about setting the values for above mentioned fields?

**Comment by mubeen** [comment]
@C0urante 

I feel we should definitely consider another task wherein we can add the feature to allow the time partitioning configuration per table which is being created. What are your thoughts?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java]: @mubeen this still needs to be addressed :)

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java]: > for the non intermediate tables I felt that we should do the look up

What purpose does the lookup serve, though? If we're creating a table from scratch, the lookup will fail and we won't get any relevant information from it. If we're updating an existing table, it won't be possible to alter the partitioning type anyway, so we could try leaving it unset in our schema update request altogether (which should be fine if BigQuery's API allows this).

With the case of timestamp field names, I think the same logic applies: if we're creating a table, we stick to the default of `DAY` time partitioning and provide the timestamp column name; if we're updating a table, we leave out partitioning information altogether and see it if remains unchanged since partitioning types for tables can't be updated after they're created anyways.

With the case of table clustering, it's possible that the same logic applies as well. The only difference I can see is that the clustering style for a table can be updated after it's created (see https://cloud.google.com/bigquery/docs/creating-clustered-tables#modifying-cluster-spec), but if we simply refrain from specifying a clustering style during a schema update, it's still possible that the current clustering style for the table will be preserved.

**Comment by C0urante** [comment]
@mubeen I think it makes sense to add support for setting the partitioning time type for tables created by the connector in the same release that adds support for writing to tables with other partitioning time types. However, I think it's fine to merge a PR for one without the other; we just have to issue a follow-up PR before cutting a release. If you can take on the work of getting the changes proposed here to a mergeable state, I'd be happy to file the follow-up PR to expose that configurability to users so that we can get this feature out without too much delay.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:358]: I think we can simplify this section:
```suggestion
    return cache.computeIfAbsent(key, k -> getBigQuery().getTable(tableId));
```

Also, why use a custom string key here instead of the `TableId`?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:203]: Nit: unnecessary empty line, should be removed.
```suggestion
```

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:205]: Think we can initialize this to `DAY` partitioning here and save a few null checks down the road.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:309]: Based on the above suggestion about initializing the partitioning type to `DAY` by default, we can take out the null check here.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:329]: Based on the above suggestion about initializing the partitioning type to DAY by default, we can take out the null check here.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:601]: Looks like this section isn't used anymore; we should revert this and all the other changes on this file if they're not needed.

**Review by C0urante — COMMENTED** [review]
Thanks @mubeen, left a few more comments. Also, can you confirm that leaving the partitioning type out during update operations works as expected?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (C0urante) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:93]: @mubeen can you please revert _**all**_ changes to this file as none appear necessary?

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/SchemaManagerTest.java
Comment (C0urante) [inline kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/SchemaManagerTest.java:50]: @mubeen can you please revert all changes to this file as none appear necessary?

Comment (mubeen) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/SchemaManagerTest.java]: Hi @C0urante 

File was reverted back to commit 26ecf5e9f8d3d524cea1dec999d10a222f74539f

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (mubeen) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java]: Hi @C0urante 

File was reverted back to commit 26ecf5e9f8d3d524cea1dec999d10a222f74539f

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (mubeen) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java]: Reworked this piece of code to use map's computeIfAbsent method along with tableId as the key as opposed to customer key.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (mubeen) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java]: Initialisation to DAY completed along with removal of the NULL checks

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (mubeen) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java]: Initialisation to DAY completed along with removal of the NULL checks

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java
Comment (mubeen) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java]: Initialisation to DAY completed along with removal of the NULL checks

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (mubeen) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java]: File has been reverted to commit 26ecf5e9f8d3d524cea1dec999d10a222f74539f

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (mubeen) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java]: File has been reverted to commit 26ecf5e9f8d3d524cea1dec999d10a222f74539f

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java]: @mubeen I've gotten around to doing some testing here and there's some good news and some bad news.

Good news: sending a table schema update request with no time partitioning specified succeeds, and doesn't remove the existing partitioning type for the table (which makes since since BigQuery doesn't support changing the partitioning type for existing tables anyways).

However, sending a table schema update request with a different time partitioning type than the one the table uses does not succeed and results in a 400 response from BigQuery.

So, I believe we need to alter the `SchemaManager::constructTableInfo` method to not include a time partitioning type at all when  a schema is being updated, and instead only include one when a new table is being created.

**Comment by C0urante** [comment]
Thanks @mubeen. I still hadn't heard from you about the question I asked in my last review:

> Also, can you confirm that leaving the partitioning type out during update operations works as expected?

so I tested for this myself. I've left my findings in https://github.com/confluentinc/kafka-connect-bigquery/pull/61#discussion_r565784712; please take a look and address this. Once that's taken care of, I think we should be close to merging; I'll still have to give the tests a look, but hopefully that should be straightforward compared to the functional parts of this PR.

**Comment by mubeen** [comment]
@C0urante 

I have made changes to determine if we are doing a table creation vs update and only applied the time partitioning to creation operation. Please review and verify my thinking.

Thanks

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:463]: @mubeen Somehow this change got reintroduced, can we please revert?

**Review by C0urante — COMMENTED** [review]
Thanks @mubeen. I've left a comment about unnecessary formatting changes on the `SchemaManager` class; please address.

The new unit tests added for the `PartitionedTableId` class look great, nice work!

One final high-level request before we merge--can we add a test to ensure that the `SchemaManager`'s logic for including/not including time partitioning types depending on whether a table is being created or updated is correct? After that, as long as the formatting comment is addressed and the new test(s) look good, I think we'll be ready to merge.

**Review by C0urante — COMMENTED** [review]
Looks good, please revert the unnecessary formatting changes and then this should be ready to merge.

**Comment by mubeen** [comment]
Hi @C0urante,

Tests added, please verify for me. Thanks

**Comment by C0urante** [comment]
@mubeen see my latest review: https://github.com/confluentinc/kafka-connect-bigquery/pull/61#pullrequestreview-579331086

**Comment by mubeen** [comment]
@C0urante 

Formatting changes completed 

Thanks

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/SchemaManagerTest.java
Comment (C0urante) [inline kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/SchemaManagerTest.java:481]: Unnecessary formatting change; please revert.

**Comment by C0urante** [comment]
Thanks @mubeen but there are still unnecessary formatting changes in this pull request. I won't be merging it until all are addressed; please check the diff on GitHub carefully to ensure that none are present.

Changes that need to be reverted:
- https://github.com/confluentinc/kafka-connect-bigquery/pull/61#discussion_r566878079
- https://github.com/confluentinc/kafka-connect-bigquery/pull/61#discussion_r567836208

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/SchemaManagerTest.java
Comment (mubeen) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/SchemaManagerTest.java:482]: Formatting changes completed, please verify. Thanks

Comment (C0urante) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/SchemaManagerTest.java:482]: This is still incorrect; see the diff in the GitHub UI.

Comment (mubeen) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/SchemaManagerTest.java]: Added the additional line back.

**Review by C0urante — APPROVED** [review]
LGTM, thanks @mubeen. Will merge pending a green Jenkins build.

**Comment by mubeen** [comment]
Thanks @C0urante, appreciate all of your assistance and patience.

**Comment by C0urante** [comment]
Thanks for the contribution @mubeen! I've filed https://github.com/confluentinc/kafka-connect-bigquery/issues/80 as a follow-up to this PR so that we can add configurable support for controlling what time partitioning type should be used for tables the connector creates. Once that work is complete, we can cut a 2.1.x feature branch and, as soon as 2.0.0 is released, release version 2.1.0 with support for the additional time partitioning types.

**Pull request merged**
